hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/cache/PublishCacheManager.java

```diff
@@ -99,26 +100,35 @@ public void init() {
 
             definitions.registerListener(events -> {
                 log.debugf("Receiving %s", events);
-                events.stream().forEach(e -> {
-                    log.debugf("Received %s", e);
-                    String tenantId = e.getTargetTenantId();
-                    String triggerId = e.getTargetId();
-                    TriggerKey triggerKey = new TriggerKey(tenantId, triggerId);
-                    publishCache.startBatch();
-                    publishDataIdsCache.startBatch();
-                    Set<String> oldDataIds = publishDataIdsCache.get(triggerKey);
-                    removePublishCache(tenantId, triggerId, oldDataIds);
-                    if (e.getType().equals(TRIGGER_CONDITION_CHANGE)) {
-                        Set<String> newDataIds = e.getDataIds();
-                        publishDataIdsCache.put(triggerKey, newDataIds);
-                        addPublishCache(tenantId, triggerId, newDataIds);
-                    }
-                    publishDataIdsCache.endBatch(true);
-                    publishCache.endBatch(true);
-                });
+                events.stream()
+                        .sorted((e1, e2) -> e1.getType() == TRIGGER_REMOVE ? 1 : -1) // process removals last
```

Comment:
Right, good catch.  I can revert the sort, it was intended just as an optimization.  I think the rest of the optimization is fine, although it too is not related to the core fix, so let me know if you think there is still an issue.